jlm/util/TarjanSCC.hpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_UTIL_TARJANSCC_HPP
+#define JLM_UTIL_TARJANSCC_HPP
+
+#include <jlm/util/common.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <stack>
+#include <vector>
+
+namespace jlm::util
+{
+
+/**
+ * Implementation of Tarjan's algorithm for finding strongly connected components in linear time.
+ * Each node is assigned an SCC, with scc indices starting at 0.
+ * An SCC has a lower index than all its predecessor SCCs, and a larger index than its successors.
+ *
+ * In addition to assigning SCCs, a partial topological ordering of nodes is returned.
+ * All nodes belonging to a given SCC are in a continuous subsequence.
+ * All nodes belonging to predecessor SCCs come earlier in the list,
+ * and all nodes belonging to successor SCCs come later.
+ *
+ * @tparam SuccessorFunctor a functor with the signature (size_t) -> iterable<size_t>
+ * @param numNodes the number of nodes. Nodes are indexed from 0 to numNodes-1
+ * @param successors an instance of the SucessorFunctor
+ * @param sccIndex output vector to be filled with the index of the SCC each node ends up in
+ * @param topologicalOrder output vector to be filled with the nodes in a weak topological order
+ * @return the number of SCCs in the graph. One more than the largest SCC index
+ */
+template<typename SuccessorFunctor>
+size_t
+FindStronglyConnectedComponents(
```

Comment:
I am quite excited about this, but I am worried whether this method is not general enough to replace other implementations:
1. https://github.com/phate/jlm/blob/f9cc3456d60c9df68b0f55c66cfb9cdf560eef2d/jlm/llvm/ir/ipgraph.cpp#L12C1-L52C2
2. https://github.com/phate/jlm/blob/f9cc3456d60c9df68b0f55c66cfb9cdf560eef2d/jlm/llvm/ir/cfg-structure.cpp#L82C1-L128C2

Reply:
I agree it could be even more general, by not using a continuous range of indices. I will try to make a light wrapper and see how it ends up.